File: src/solutions/2025/01/zig/main.zig
const std = @import("std");
const aoc = @import("aoc");

const options = aoc.SolutionOptions.init();

fn solve(input: []const u8, is_test: bool, p2: bool) []const u8 {
    _ = is_test;
    var dial: i32 = 50;
    var zero: i32 = 0;

    var lines = std.mem.splitSequence(u8, input, "\n");
    while (lines.next()) |line| {
        const newDial = dial + (std.fmt.parseInt(i32, line[1..], 10) catch 0) * @as(i32, if (line[0] == 'R') 1 else -1);

        if (p2) {
            zero += @divFloor(if (newDial < 0) newDial * -1 else newDial, 100) + @as(i32, if (dial != 0 and newDial <= 0) 1 else 0);
            dial = @mod(newDial, 100);
        } else {
            dial = newDial;
            if (@mod(dial, 100) == 0) {
                zero += 1;
            }
        }
    }

    const buf = std.heap.c_allocator.alloc(u8, 42) catch unreachable;
    const str = std.fmt.bufPrint(buf, "{}", .{zero}) catch "";
    return str;
}

fn part1(input: []const u8, is_test: bool) []const u8 {
    return solve(input, is_test, false);
}

fn part2(input: []const u8, is_test: bool) []const u8 {
    return solve(input, is_test, true);
}

pub fn main() !void {
    try aoc.run(part1, part2, options);
}

File: src/solutions/2025/01/rust/main.rs
extern crate aoc_core;
use std::{env, iter::FromIterator};

fn get_options() -> aoc_core::options::SolutionOptions {
   aoc_core::options::SolutionOptions {
      ..Default::default()
   }
}

fn modulo(x: i32, y: i32) -> i32 {
   let r = x % y;
   if r < 0 {
      r + y
   } else {
      r
   }
}

fn solve(input: &str, _is_test: bool, p2: bool) -> String {
   let mut dial = 50;
   input
      .lines()
      .map(|line| {
         let new_dial = dial
            + line[1..].parse::<i32>().unwrap()
               * if line.chars().nth(0).unwrap() == 'R' {
                  1
               } else {
                  -1
               };

         if p2 {
            let zero = new_dial.abs() / 100 + if dial != 0 && new_dial <= 0 { 1 } else { 0 };
            dial = modulo(new_dial, 100);
            return zero;
         } else {
            dial = new_dial;
            if dial % 100 == 0 {
               return 1;
            }
         }
         return 0;
      })
      .sum::<i32>()
      .to_string()
}

fn part_1(input: &str, is_test: bool) -> String {
   solve(input, is_test, false)
}

fn part_2(input: &str, is_test: bool) -> String {
   solve(input, is_test, true)
}

fn main() {
   let args = Vec::from_iter(env::args());
   aoc_core::run(
      args.iter().map(|x| x.as_ref()).collect::<Vec<_>>(),
      part_1,
      part_2,
      get_options(),
   );
}
